mining/src/mempool/model/frontier.rs

```diff
@@ -0,0 +1,391 @@
+    ///        this by computing the prefix weight up to this top item (exclusive) and then continue the sampling process over
+    ///        the narrowed space. This process is repeated until acquiring the desired mass.  
+    ///     4. Numerical stability. Naively, one would simply subtract `total_weight -= top.weight` in order to narrow the sampling
+    ///        space. However, if `top.weight` is much larger than the remaining weight, the above f64 subtraction will yield a number
+    ///        close or equal to zero. We fix this by implementing a `log(n)` prefix weight operation.
+    ///     5. Q. Why not just use u64 weights?
+    ///        A. The current weight calculation is `feerate^alpha` with `alpha=3`. Using u64 would mean that the feerate space
+    ///           is limited to a range of size `(2^64)^(1/3) = ~2^21 = ~2M`. Already with current usages, the feerate can vary
+    ///           from `~1/50` (2000 sompi for a transaction with 100K storage mass), to `5M` (100 KAS fee for a transaction with
+    ///           2000 mass = 100·100_000_000/2000), resulting in a range of size 250M (`5M/(1/50)`).
+    ///           By using floating point arithmetics we gain the adjustment of the probability space to the accuracy level required for
+    ///           current samples. And if the space is highly biased, the repeated elimination of top items and the prefix weight computation
+    ///           will readjust it.
+    pub fn sample_inplace<R>(&self, rng: &mut R, policy: &Policy) -> SequenceSelectorInput
+    where
+        R: Rng + ?Sized,
+    {
+        debug_assert!(!self.search_tree.is_empty(), "expected to be called only if not empty");
+
+        // Sample 20% more than the hard limit in order to allow the SequenceSelector to
+        // compensate for consensus rejections.
+        // Note: this is a soft limit which is why the loop below might pass it if the
+        //       next sampled transaction happens to cross the bound
+        let desired_mass = (policy.max_block_mass as f64 * MASS_LIMIT_FACTOR) as u64;
+
+        let mut distr = Uniform::new(0f64, self.total_weight());
+        let mut down_iter = self.search_tree.descending_iter();
+        let mut top = down_iter.next().unwrap();
+        let mut cache = HashSet::new();
+        let mut sequence = SequenceSelectorInput::default();
+        let mut total_selected_mass: u64 = 0;
+        let mut _collisions = 0;
+
+        // The sampling process is converging so the cache will eventually hold all entries, which guarantees loop exit
+        'outer: while cache.len() < self.search_tree.len() && total_selected_mass <= desired_mass {
+            let query = distr.sample(rng);
+            let item = {
+                let mut item = self.search_tree.search(query);
+                while !cache.insert(item.tx.id()) {
+                    _collisions += 1;
+                    if top == item {
```

Comment:
I think the reason you chose this as a method for reducing the sampling collisions (by reducing the search space) is because the `top` is the item that has the likeliest probability of being chosen, given the weight. Is that right?

Reply:
Correct. See my other [comment](https://github.com/kaspanet/rusty-kaspa/pull/513#discussion_r1712984709) form more on the subject

Reply:
Also, it would be challenging to reduce the space by dropping items from the middle of the tree (higher complexity + not supported by the visitor pattern I used).